price-estimator/src/filter.rs

```diff
@@ -0,0 +1,218 @@
+use crate::orderbook::Orderbook;
+use serde::Serialize;
+use serde_with::rust::display_fromstr;
+use std::num::ParseIntError;
+use std::sync::Arc;
+use warp::{Filter, Rejection};
+
+/// Validate a request of the form
+/// `/markets/<baseTokenId>-<quoteTokenId>/estimated-buy-amount/<sellAmountInQuoteToken>`
+/// and answer it.
+pub fn estimated_buy_amount<T: Send + Sync + 'static>(
+    orderbook: Arc<Orderbook<T>>,
+    price_rounding_buffer: f64,
+) -> impl Filter<Extract = (String,), Error = Rejection> + Clone {
+    estimated_buy_amount_filter().and_then({
+        move |token_pair, sell_amount_in_quote| {
+            let orderbook = orderbook.clone();
+            async move {
+                let orderbook = orderbook.get_reduced_orderbook().await;
+                let result = estimate_buy_amount(
+                    token_pair,
+                    sell_amount_in_quote,
+                    price_rounding_buffer,
+                    orderbook,
+                );
+                // The compiler cannot infer the error type because we never return an error.
+                Result::<_, Rejection>::Ok(result)
+            }
+        }
+    })
+}
+
+fn estimated_buy_amount_filter(
+) -> impl Filter<Extract = (TokenPair, u128), Error = Rejection> + Copy {
+    warp::path!("markets" / TokenPair / "estimated-buy-amount" / u128)
+}
+
+fn estimate_buy_amount(
+    token_pair: TokenPair,
+    sell_amount_in_quote: u128,
+    price_rounding_buffer: f64,
+    orderbook: pricegraph::Orderbook,
+) -> String {
+    let buy_amount_in_base = crate::estimate_buy_amount::estimate_buy_amount(
+        token_pair,
+        sell_amount_in_quote as f64,
+        price_rounding_buffer,
+        orderbook,
+    )
+    .unwrap_or(0.0) as u128;
```

Comment:
How did we decide on this default of zero? Is this what we plan to return when the estimation fails? Could this be interpreted as misleading to the end user?

Reply:
> Is this what we plan to return when the estimation fails?

Yes. The pricegraph documents this

>   Returns `None` if the order cannot be filled because the token pair is not connected, or the connection cannot support enough capacity to fully fill the specified volume.

I agree that this should be an error instead of 0 but this is the same behavior as the nodejs price estimator. We should change it in another PR.